lib/timber/transports/http.ex

```diff
@@ -0,0 +1,189 @@
+defmodule Timber.Transports.HTTP do
+  @moduledoc """
+  An efficient HTTP transport that buffers and delivers log messages over HTTP to the
+  Timber API. It uses batching, keep-alive connections, and msgpack to deliver logs with
+  high-throughput and little overhead.
+
+  The HTTP transport functions differently than a traditional buffer in that it is designed for
+  batch delivery. This means messages are buffered by default and flushed on an internval.
+  If the buffer size exceeds `:max_buffer_size` before the next intervaled flush, the buffer
+  will be immediately flushed.
+
+  All outgoing requests are made asynchronously. If a second request is made while the
+  previous (first) request is still being processed, then the transport will enter
+  synchronous mode, waiting for a response before proceeding with the request.
+
+  ## Configuration
+
+  ### Custom HTTP client
+
+  By default, hackney is used via `Timber.Transports.HTTP.HackneyClient`. You can define your
+  own custom HTTP client by adhering to the `Timber.Transports.HTTP.Client` behaviour. Afterwards,
+  you must specify your client in the configuration:
+
+  ```
+  config :timber, :http_transport, http_client: MyHTTPClient
+  ```
+  """
+
+  @behaviour Timber.Transport
+
+  alias Timber.LogEntry
+
+  @typep t :: %__MODULE__{
+    api_key: String.t,
+    buffer_size: non_neg_integer,
+    buffer: [] | [IO.chardata],
+    flush_interval: non_neg_integer,
+    max_buffer_size: pos_integer,
+    ref: reference
+  }
+
+  @content_type "application/msgpack"
+  @default_max_buffer_size 5000 # 5000 log line should be well below 5mb
+  @default_flush_interval 1000
+  @default_http_client __MODULE__.HackneyClient
+  @url "https://api.timber.io/frames"
+
+  defstruct api_key: nil,
+            buffer_size: 0,
+            buffer: [],
+            flush_interval: @default_flush_interval,
+            max_buffer_size: @default_max_buffer_size,
+            ref: nil
+
+  @doc false
+  @spec init() :: {:ok, t}
+  def init() do
+    config = Keyword.put(config(), :api_key, Timber.Config.api_key())
+
+    case configure(config, %__MODULE__{}) do
+      {:ok, state} ->
+        flusher(state)
+        {:ok, state}
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  @doc false
+  @spec configure(Keyword.t, t) :: {:ok, t}
+  def configure(options, %{api_key: current_api_key} = state) do
+    api_key = Keyword.get(options, :api_key, current_api_key)
+    max_buffer_size = Keyword.get(options, :max_buffer_size, @default_max_buffer_size)
+    new_state = %{ state | api_key: api_key, max_buffer_size: max_buffer_size }
+
+    if api_key == nil do
+      {:error, :no_api_key}
+    else
+      {:ok, new_state}
+    end
+  end
+
+  @doc false
+  @spec write(LogEntry.t, t) :: {:ok, t}
+  def write(log_entry, state) do
+    # Write to the buffer immediately because we want to batch lines and send
+    # them on an interval.
+    state = write_buffer(log_entry, state)
```

Comment:
I'm not following, this would not allow the buffer to grow beyond the `max_buffer_size`. Yes, it checks the size after buffering, but that only means we're account for the new member when checking if the buffer is full. The only scenario where the buffer would grow endlessly is if the `max_buffer_size` was set to <= 0. But I'll add a `>=` for safety reasons.